service-base/src/main/java/org/eclipse/hono/service/amqp/RequestResponseEndpoint.java

```diff
@@ -237,54 +296,182 @@ protected final void handleMessage(final ProtonConnection con, final ProtonRecei
+
+        final MessageConsumer<JsonObject> replyConsumer = vertx.eventBus().consumer(replyTo,
+                message -> {
+                    // TODO check for correct session here...?
+                    if (logger.isTraceEnabled()) {
+                        logger.trace("forwarding reply to client [{}]: {}", sender.getName(),
+                                message.body().encodePrettily());
+                    }
+                    final EventBusMessage response = EventBusMessage.fromJson(message.body());
+                    filterResponse(Constants.getClientPrincipal(con), response)
+                            .recover(t -> {
+                                final int status = ServiceInvocationException.extractStatusCode(t);
+                                return Future.succeededFuture(response.getResponse(status));
+                            })
+                            .map(filteredResponse -> {
+                                final Message amqpReply = getAmqpReply(filteredResponse);
+                                sender.send(amqpReply);
+                                flowCreditToRequestor(replyTo);
+                                return null;
+                            })
+                            .recover(t -> {
+                                flowCreditToRequestor(replyTo);
+                                return Future.failedFuture(t);
+                            });
+                });
+
+        // register this consumer and replyTo address with this connection
+
+        registerConsumerForConnection(con, replyTo, replyConsumer);
+
+        sender.setQoS(ProtonQoS.AT_LEAST_ONCE);
+
+        HonoProtonHelper.setCloseHandler(sender, senderClosed -> {
+            logger.debug("client [{}] closed sender link, removing associated event bus consumer [{}]",
+                    sender.getName(), replyConsumer.address());
+
+            deallocateReceiverForReplyTo(replyTo);
+            unregisterConsumerForConnection(con, replyTo, replyConsumer);
+
+            if (senderClosed.succeeded()) {
+                senderClosed.result().close();
+            }
+        });
+
+        sender.open();
+
+    }
+
+    @Override
+    public void onConnectionClosed(final ProtonConnection connection) {
+
+        Objects.requireNonNull(connection);
+
+        unregisterAllConsumersForConnection(connection);
+        deallocateAllReceiversForConnection(connection);
+    }
+
+    private void allocateReceiverForReplyTo(final String replyTo, final ProtonReceiver receiver) {
+
+        final ProtonReceiver oldReceiver = replyToReceiverMap.put(replyTo, receiver);
+
+        if (oldReceiver == null || oldReceiver == receiver) {
+            logger.debug("Allocated receiver [{}] for replies to [{}]", receiver, replyTo);
+        } else {
+            logger.info("Allocated receiver [{}] for replies to [{}] - Had existing receiver: [{}]", receiver, replyTo,
```

Comment:
this means that we have multiple clients using the same reply-to address, right?

Reply:
Yes.